File: csharp/hamming/Hamming.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Hamming
{
    public static int Distance(string first, string second)
    {
        if (first.Length != second.Length) {
            throw new ArgumentException();
        }
        return first.ToCharArray()
            .Zip(second.ToCharArray(), CountIfDifferent)
            .Sum();
    }

    private static int CountIfDifferent(char a, char b) => a != b ? 1 : 0;
}

File: fsharp/hamming/Hamming.fs
module Hamming

let countIfDifferent (a, b): int = 
    if a <> b then 1 else 0

let distance (s1: string) (s2: string): int option = 
    if s1.Length <> s2.Length then None else
        Seq.zip s1 s2
        |> Seq.sumBy countIfDifferent
        |> Some

// let different (a, b) = 
//     if a <> b then Some (a, b) else None

// let distance (s1: string) (s2: string): int option = 
//     if s1.Length <> s2.Length then None else
//         Seq.zip s1 s2
//         |> Seq.map different
//         |> Seq.choose id
//         |> Seq.length
//         |> Some